x/evm/keeper/grpc_query.go

```diff
@@ -364,31 +365,23 @@ func (k Keeper) EstimateGas(c context.Context, req *types.EthCallRequest) (*type
 		return len(rsp.VmError) > 0, rsp, nil
 	}
 
-	// Execute the binary search and hone in on an executable gas limit
-	hi, err = types.BinSearch(lo, hi, executable)
+	// Reject the transaction as invalid if it still fails at the highest allowance
+	failed, result, err := executable(hi)
 	if err != nil {
 		return nil, err
 	}
 
-	// Reject the transaction as invalid if it still fails at the highest allowance
-	if hi == cap {
-		failed, result, err := executable(hi)
-		if err != nil {
-			return nil, err
-		}
-
-		if failed {
-			if result != nil && result.VmError != vm.ErrOutOfGas.Error() {
-				if result.VmError == vm.ErrExecutionReverted.Error() {
-					return nil, types.NewExecErrorWithReason(result.Ret)
-				}
-				return nil, errors.New(result.VmError)
+	if failed {
+		if result != nil && result.VmError != vm.ErrOutOfGas.Error() {
+			if result.VmError == vm.ErrExecutionReverted.Error() {
+				return nil, types.NewExecErrorWithReason(result.Ret)
 			}
-			// Otherwise, the specified gas cap is too low
-			return nil, fmt.Errorf("gas required exceeds allowance (%d)", cap)
+			return nil, errors.New(result.VmError)
 		}
+		// Otherwise, the specified gas cap is too low
+		return nil, fmt.Errorf("gas required exceeds allowance (%d)", hi)
 	}
-	return &types.EstimateGasResponse{Gas: hi}, nil
+	return &types.EstimateGasResponse{Gas: uint64(float64(result.GasUsed) * 1.1)}, nil
```

Comment:
GasUsed*1.1 may be larger than req.GasCap, and it may be insufficient if there are over 10% gas refunds in the execution?